libs/constrained-generators/src/Constrained/Base.hs

```diff
@@ -6122,3 +6118,131 @@ nubOrdMemberSpec message xs =
         , "The input is the empty list."
         ]
     )
+
+-- ==============================================================================
+-- Experimental changes to conformsToSpecM
+
+runTermE :: Env -> Term fn a -> Either (NE.NonEmpty String) a
+runTermE env = \case
+  Lit a -> Right a
+  V v -> case lookupEnv env v of
+    Just a -> Right a
+    Nothing -> Left (pure ("Couldn't find " ++ show v ++ " in " ++ show env))
+  App f ts -> do
+    vs <- mapMList (fmap Identity . runTermE env) ts
+    pure $ uncurryList_ runIdentity (sem f) vs
+
+checkPredsE ::
+  FunctionLike fn =>
+  NE.NonEmpty String ->
+  Env ->
+  [Pred fn] ->
+  Maybe (NE.NonEmpty String)
+checkPredsE msgs env ps =
+  case catMaybes (fmap (checkPredE env msgs) ps) of
+    [] -> Nothing
+    (x : xs) -> Just (NE.nub (foldl (<>) x xs))
+
+checkPredE ::
+  forall fn.
+  FunctionLike fn =>
+  Env -> NE.NonEmpty String -> Pred fn -> Maybe (NE.NonEmpty String)
+checkPredE env msgs = \case
+  Monitor {} -> Nothing
+  Subst x t p -> checkPredE env msgs $ substitutePred x t p
+  Assert t -> case runTermE env t of
+    Right True -> Nothing
+    Right False ->
+      Just
+        (msgs <> pure ("Assert " ++ show t ++ " returns False") <> pure ("\nenv=\n" ++ show (pretty env)))
+    Left es -> Just (msgs <> es)
+  GenHint {} -> Nothing
+  p@(Reifies t' t f) ->
+    case runTermE env t of
+      Left es -> Just (msgs <> NE.fromList ["checkPredE: Reification fails", "  " ++ show p] <> es)
+      Right val -> case runTermE env t' of
+        Left es -> Just (msgs <> NE.fromList ["checkPredE: Reification fails", "  " ++ show p] <> es)
+        Right val' ->
+          if f val == val'
+            then Nothing
+            else
+              Just
+                ( msgs
+                    <> NE.fromList
+                      [ "checkPredE: Reification doesn't match up"
+                      , "  " ++ show p
+                      , show (f val) ++ " /= " ++ show val'
+                      ]
+                )
+  ForAll t (x :-> p) -> case runTermE env t of
+    Left es -> Just $ (msgs <> NE.fromList ["checkPredE: ForAll fails to run."] <> es)
+    Right set ->
+      let answers =
+            catMaybes
+              [ checkPredE env' (pure "Some items in ForAll fail") p
+              | v <- forAllToList set
+              , let env' = extendEnv x v env
+              ]
+       in case answers of
+            [] -> Nothing
+            (x : xs) -> Just (msgs <> NE.nub (foldr (<>) x xs))
```

Comment:
My suggestion about `sconcat` applies here too

Reply:
fixed it and 2 other places as well.